analyzer/metadata-plugin/src/main/java/eu/fasten/analyzer/metadataplugin/MetadataDBExtension.java

```diff
@@ -0,0 +1,335 @@
+            callgraph =  getExtendedRevisionCallGraph(forge, consumedJson);
+        } catch (JSONException e) {
+            logger.error("Error parsing JSON callgraph for '"
+                    + Paths.get(path).getFileName() + "'", e);
+            processedRecord = false;
+            setPluginError(e);
+            return;
+        }
+
+        var revision = setOutputPath(callgraph);
+
+        int transactionRestartCount = 0;
+        do {
+            setPluginError(null);
+            try {
+                var metadataDao = new MetadataDao(getDBConnection());
+                getDBConnection().transaction(transaction -> {
+                    // Start transaction
+                    metadataDao.setContext(DSL.using(transaction));
+                    long id;
+                    try {
+                        id = saveToDatabase(callgraph, metadataDao);
+                    } catch (RuntimeException e) {
+                        processedRecord = false;
+                        logger.error("Error saving to the database: '" + revision + "'", e);
+                        setPluginError(e);
+                        if (e instanceof DataAccessException) {
+                            // Database connection error
+                            if (e.getCause() instanceof BatchUpdateException) {
+                                var exception = ((BatchUpdateException) e.getCause())
+                                        .getNextException();
+                                setPluginError(exception);
+                            }
+                            logger.info("Restarting transaction for '" + revision + "'");
+                            // It could be a deadlock, so restart transaction
+                            restartTransaction = true;
+                        } else {
+                            restartTransaction = false;
+                        }
+                        throw e;
+                    }
+                    if (getPluginError() == null) {
+                        processedRecord = true;
+                        restartTransaction = false;
+                        logger.info("Saved the '" + revision + "' callgraph metadata "
+                                + "to the database with package version ID = " + id);
+                    }
+                });
+            } catch (Exception expected) {
+            }
+            transactionRestartCount++;
+        } while (restartTransaction && !processedRecord
+                && transactionRestartCount < Constants.transactionRestartLimit);
+    }
+
+    @Override
+    public Optional<String> produce() {
+        if (gidGraph == null) {
+            return Optional.empty();
+        } else {
+            return Optional.of(gidGraph.toJSONString());
+        }
+    }
+
+    @Override
+    public String getOutputPath() {
+        return outputPath;
+    }
+
+    // FIXME Add comments, decouple logic, make it language agnostic
```

Comment:
I think the method is a bit incorrect. The main point of the method is to set output path to be `/forge/first-letter-of-group/group_artifact_version.json`. The rest is not really needed. So basically everything that should be there is actually between lines 189 and 196, but group and artifact should be interchanged and version should be added when setting outputPath. This way it will also be language-agnostic.

Reply:
Could you please fix it?

Reply:
Sure, I will fix it on Monday.

Reply:
@StefanosChaliasos it is done